### ISSUE TITLE
Initialization of Decimal without str?

### ISSUE THREAD

**Issue by ChristopherRabotin** [dnsl48/fraction#36]
Hi there,

I was wondering if there were a way to initialize a Decimal without passing through a string formatting?

Looking at the flamegraph of a library which does literally thousands of computations using decimal, the conversion of an `f64` into a `GenericDecimal<u128, u16>` takes 4.16% of the computation time (cf. attachment).

Hence, I was wondering what I could increase the performance. Thanks

![Screenshot_20201207_195935](https://user-images.githubusercontent.com/4823784/101433107-5c1d3f00-38c7-11eb-9616-8d4a420004e1.png)

**Comment by dnsl48** [comment]
Hi @ChristopherRabotin,

Our Decimal implementation relies internally on the Fraction type, so the most efficient way would be to figure out how to convert `f64` into a `Fraction` and then initialise the `Decimal` via its constructor, e.g. `GenericDecimal(fraction, fixed_precision)`.
The most flexible (and easy) way I have found initially was to convert f32/64 into a string and then convert it into a fraction. It's implemented here - https://github.com/dnsl48/fraction/blob/master/src/fraction/mod.rs#L453
I considered some ways to optimise that via working with float binary representation, but those Rust APIs weren't stable back then and later I didn't find time to look into it again. Perhaps that could be a way to optimise such conversion.
If you come up with any other ideas, or a more efficient implementation, please let us know. PRs are very welcome too! :)

**Comment by ChristopherRabotin** [comment]
Awesome, thanks for the link to the specific part of the code. I'll see if
I can find another approach (absolutely no guarantees, you obviously have
thought about the problem much more than I have!)

Christopher Rabotin.

**Comment by ChristopherRabotin** [comment]
This algorithm seems to work: https://math.stackexchange.com/a/1049723/17452 .

Here is demo code in Python (I'll try converting that to Rust with the num traits soon):

```py
In [6]: def to_frac(f): 
   ...:     p = 0 
   ...:     while True: 
   ...:         if abs(round(f) - f) < 2e-16: 
   ...:             break 
   ...:         p += 1 
   ...:         f *= 10**p 
   ...:     # end while 
   ...:     g = gcd(int(f), 10**p) 
   ...:     print(g) 
   ...:     num = f/g 
   ...:     den = (10**p)/g 
   ...:     print(num, den) 
   ...:  
(...)
In [10]: to_frac(pi)                                                                                                                                                                                                                                                                                                                               
1
3141592653589793.0 100000.0

In [11]: to_frac(0.5)                                                                                                                                                                                                                                                                                                                              
5
1.0 2.0

In [12]: to_frac(128959848.1457456)                                                                                                                                                                                                                                                                                                                
16
8.0599905091091e+16 625.0

In [13]
```

**Comment by dnsl48** [comment]
Yes, that looks like the most straightforward way to do it, but it has a flaw.

> // Multiply by the precision
            new_src *= 10_f64.powi(p_i as i32);

Here might be an overflow for huge numbers, so we can only do that if we're sure there's enough space in f64 for this in-place computation.

**Comment by dnsl48** [comment]
> Here might be an overflow for huge numbers, so we can only do that if we're sure there's enough space in f64 for this in-place computation.

I guess that's actually an edge case, so we may fallback to the _stringify_ version if that happens. That could allow us to be more efficient in general and still handle big numbers gracefully.

**Comment by ChristopherRabotin** [comment]
Cool, I've got this implemented, and criterion says it's a pretty significant improvement: 75% speed up.

```
     Running target/release/deps/bench_fraction-2f3e4b9d43259cec
Gnuplot not found, using plotters backend
Decimal u128/u16 init   time:   [112.55 ns 112.95 ns 113.36 ns]                                  
                        change: [-77.847% -77.728% -77.609%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

Decimal i64/u16 init    time:   [107.26 ns 107.83 ns 108.49 ns]                                 
                        change: [-76.184% -75.792% -75.481%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

```

Just gotta clean things up and I'll open a PR. I'll bump it to version 0.7.1 if that's cool, although maybe it's worth using 0.8?

**Comment by dnsl48** [comment]
That appears very promising!

> I'll bump it to version 0.7.1 if that's cool, although maybe it's worth using 0.8?

I would suggest to do it as `0.7.1`. This does not break any backward compatibility. Adding new API should be fine because we're still within the version `0.*`, so semantic versioning shouldn't prevent us from doing that.

**Comment by dnsl48** [comment]
Released with `0.8.0`.
Thank you very much for the contribution!
